src/tactic/fin_cases.lean

```diff
@@ -3,38 +3,75 @@ Copyright (c) 2018 Scott Morrison. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
 Author: Scott Morrison
 
-Case bashing on `fin n`, for explicit numerals `n`.
+Case bashing:
+* on `x ∈ A`, for `A : finset α` or `A : list α`, or
+* on `x : A`, with `[fintype A]`.
 -/
-import data.fin
+import data.fintype
+import tactic.norm_num
 
 namespace tactic
 open lean.parser
 open interactive interactive.types expr
+open conv.interactive
 
-meta def fin_cases_at (e : expr) : tactic unit :=
-do `(fin %%n) ← infer_type e,
-   n ← eval_expr ℕ n,
-   [(_, [val, bd], _)] ← cases_core e [`val, `bd],
-   -- We now call `cases val` n times, rotating the generated goals out of the way.
-   iterate_at_most n (do val ← get_local `val, cases val, rotate 1),
-   -- We've got an absurd hypothesis `bd`, but it is messy: it looks like
-   -- `nat.succ (... (nat.succ val)) < n`
-   -- So we rewrite it as `bd : val + 1 + ... + 1 < n`, and use `dec_trivial` to kill it.
-   ss ← mk_const `nat.succ_eq_add_one,
-   bd ← get_local `bd,
-   (list.range n).mfoldl (λ bd _, do rewrite_hyp ss bd) bd,
-   to_expr ``(absurd %%bd dec_trivial) >>= exact,
-   -- We put the goals back in order, and clear the `bd` hypotheses.
-   iterate_exactly n (do rotate_right 1,
-                         try `[rw [fin.mk_val]],
-                         try (get_local `bd >>= clear))
+/-- Checks that the expression looks like `x ∈ A` for `A : finset α` or `A : list α`,
+    and returns the type α. -/
+meta def guard_mem_fin (e : expr) : tactic expr :=
+do t ← infer_type e,
+   match t with
+   | `(%%x ∈ %%A) :=
+      do t ← infer_type A,
+        match t with
+        | `(finset %%α) := return α
+        | `(list %%α)   := return α
+        | _ := failed
+        end
+   | `(@list.mem %%α %%x %%A) := return α
+   | _ := failed
+   end
+
+meta def fin_cases_at : expr → tactic unit
+| e :=
+focus1 $
+-- Deal with `x ∈ A` hypotheses first:
+(do ty ← guard_mem_fin e,
+   numeric ← option.is_some <$> try_core (unify ty `(ℕ) <|> unify ty `(ℤ) <|> unify ty `(ℚ)),
+   result ← cases_core e,
+   match result with
+   -- We have a goal with an equation `s`, and a second goal with a smaller `e : x ∈ _`.
+   | [(_, [s], _), (_, [e], _)] :=
+     do let sn := local_pp_name s,
+        if numeric then
```

Comment:
This can be written more cleanly as `when numeric (...)`

Reply:
Fixed.